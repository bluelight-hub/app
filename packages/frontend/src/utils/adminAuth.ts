import { BackendApi } from '@/api/api';

/**
 * Verifiziert ob der aktuelle Benutzer Admin-Rechte hat
 *
 * Diese Funktion prüft das Admin-Cookie durch einen API-Call zum Backend.
 * Das Cookie wird automatisch mitgesendet (credentials: 'include').
 *
 * @returns Promise<boolean> - true wenn Admin-Token gültig, false sonst
 */
export async function verifyAdmin(): Promise<boolean> {
  try {
    const api = new BackendApi();
    // Der verify Endpoint gibt 200 zurück wenn das Admin-Cookie gültig ist
    await api.auth().authControllerVerifyAdminToken();
    return true;
  } catch (_error) {
    // Bei 401 oder anderen Fehlern ist das Token ungültig
    return false;
  }
}
