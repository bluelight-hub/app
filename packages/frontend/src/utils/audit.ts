import React from 'react';
import { logger } from './logger';
import { useAuthStore } from '../stores/useAuthStore';

/**
 * Audit-Aktionstypen
 * TODO: Replace with generated types from @bluelight-hub/shared/client/models once available
 */
export enum AuditActionType {
  CREATE = 'CREATE',
  READ = 'READ',
  UPDATE = 'UPDATE',
  DELETE = 'DELETE',
  LOGIN = 'LOGIN',
  LOGOUT = 'LOGOUT',
  ERROR = 'ERROR',
  SECURITY = 'SECURITY',
  SYSTEM = 'SYSTEM'
}

/**
 * Audit-Schweregrade
 * TODO: Replace with generated types from @bluelight-hub/shared/client/models once available
 */
export enum AuditSeverity {
  LOW = 'LOW',
  MEDIUM = 'MEDIUM',
  HIGH = 'HIGH',
  CRITICAL = 'CRITICAL'
}

/**
 * Kontext-Informationen für Audit-Logs
 */
export interface AuditContext {
  action: string;
  resource: string;
  resourceId?: string;
  actionType: AuditActionType;
  severity?: AuditSeverity;
  oldValues?: Record<string, any>;
  newValues?: Record<string, any>;
  metadata?: Record<string, any>;
  httpMethod?: string;
  httpPath?: string;
  errorMessage?: string;
  affectedFields?: string[];
  sensitiveData?: boolean;
  compliance?: string[];
}

/**
 * Audit-Logger für Frontend-Aktionen
 * Erfasst Benutzeraktionen und sendet sie an das Backend-Audit-System
 */
class AuditLogger {
  private static instance: AuditLogger;
  private batchQueue: AuditContext[] = [];
  private batchTimer: NodeJS.Timeout | null = null;
  private readonly BATCH_SIZE = 10;
  private readonly BATCH_DELAY = 5000; // 5 seconds

  private constructor() {
    // Initialize with window unload handler to flush remaining logs
    if (typeof window !== 'undefined') {
      window.addEventListener('beforeunload', () => {
        this.flushBatch();
      });
    }
  }

  public static getInstance(): AuditLogger {
    if (!AuditLogger.instance) {
      AuditLogger.instance = new AuditLogger();
    }
    return AuditLogger.instance;
  }

  /**
   * Protokolliert eine Benutzeraktion
   */
  public async log(context: AuditContext): Promise<void> {
    try {
      const authState = useAuthStore.getState();
      const enrichedContext = this.enrichContext(context, authState);

      // Add to batch queue
      this.batchQueue.push(enrichedContext);

      // Check if we should send immediately or batch
      if (this.shouldSendImmediately(context)) {
        await this.flushBatch();
      } else if (this.batchQueue.length >= this.BATCH_SIZE) {
        await this.flushBatch();
      } else {
        this.scheduleBatch();
      }
    } catch (error) {
      logger.error('Failed to log audit event', {
        error: error instanceof Error ? error.message : String(error),
        context,
      });
    }
  }

  /**
   * Protokolliert eine erfolgreiche Aktion
   */
  public async logSuccess(
    action: string,
    resource: string,
    details?: Partial<AuditContext>,
  ): Promise<void> {
    await this.log({
      action,
      resource,
      actionType: details?.actionType || AuditActionType.READ,
      severity: details?.severity || AuditSeverity.LOW,
      ...details,
    });
  }

  /**
   * Protokolliert einen Fehler
   */
  public async logError(
    action: string,
    resource: string,
    error: Error | string,
    details?: Partial<AuditContext>,
  ): Promise<void> {
    const errorMessage = error instanceof Error ? error.message : error;
    
    await this.log({
      action,
      resource,
      actionType: details?.actionType || AuditActionType.ERROR,
      severity: details?.severity || AuditSeverity.HIGH,
      errorMessage,
      ...details,
    });
  }

  /**
   * Protokolliert eine Sicherheitsaktion
   */
  public async logSecurity(
    action: string,
    resource: string,
    details?: Partial<AuditContext>,
  ): Promise<void> {
    await this.log({
      action,
      resource,
      actionType: AuditActionType.SECURITY,
      severity: details?.severity || AuditSeverity.HIGH,
      sensitiveData: true,
      ...details,
    });
  }

  /**
   * Protokolliert eine Datenänderung
   */
  public async logDataChange(
    action: string,
    resource: string,
    resourceId: string,
    oldValues: Record<string, any>,
    newValues: Record<string, any>,
    details?: Partial<AuditContext>,
  ): Promise<void> {
    const changedFields = Object.keys(newValues).filter(
      key => JSON.stringify(oldValues[key]) !== JSON.stringify(newValues[key])
    );

    await this.log({
      action,
      resource,
      resourceId,
      actionType: AuditActionType.UPDATE,
      severity: details?.severity || AuditSeverity.MEDIUM,
      oldValues,
      newValues,
      affectedFields: changedFields,
      ...details,
    });
  }

  /**
   * Enriches the audit context with user and system information
   */
  private enrichContext(context: AuditContext, authState: any): AuditContext {
    const userAgent = typeof navigator !== 'undefined' ? navigator.userAgent : 'Unknown';
    
    return {
      ...context,
      metadata: {
        ...context.metadata,
        userAgent,
        timestamp: new Date().toISOString(),
        clientVersion: import.meta.env.VITE_APP_VERSION || 'unknown',
        userId: authState.user?.id,
        userEmail: authState.user?.email,
        userRole: authState.user?.roles?.[0],
        sessionId: authState.sessionId,
      },
    };
  }

  /**
   * Determines if an audit log should be sent immediately
   */
  private shouldSendImmediately(context: AuditContext): boolean {
    return (
      context.severity === AuditSeverity.CRITICAL ||
      context.actionType === AuditActionType.SECURITY ||
      context.actionType === AuditActionType.ERROR ||
      context.sensitiveData === true
    );
  }

  /**
   * Schedules batch processing
   */
  private scheduleBatch(): void {
    if (this.batchTimer) {
      return;
    }

    this.batchTimer = setTimeout(() => {
      this.flushBatch();
    }, this.BATCH_DELAY);
  }

  /**
   * Flushes the batch queue
   */
  private async flushBatch(): Promise<void> {
    if (this.batchQueue.length === 0) {
      return;
    }

    const logsToSend = [...this.batchQueue];
    this.batchQueue = [];

    if (this.batchTimer) {
      clearTimeout(this.batchTimer);
      this.batchTimer = null;
    }

    try {
      // Send logs to backend
      // Note: We need to implement the batch endpoint in the API
      for (const log of logsToSend) {
        await this.sendLog(log);
      }
    } catch (error) {
      logger.error('Failed to flush audit log batch', {
        error: error instanceof Error ? error.message : String(error),
        batchSize: logsToSend.length,
      });
      
      // Re-add failed logs to queue for retry
      this.batchQueue.unshift(...logsToSend);
    }
  }

  /**
   * Sends a single log entry to the backend
   */
  private async sendLog(context: AuditContext): Promise<void> {
    try {
      // TODO: Replace with actual API call once AuditApi is generated
      // await api.audit.create({
      //   actionType: context.actionType,
      //   action: context.action,
      //   resource: context.resource,
      //   resourceId: context.resourceId,
      //   ...context,
      // });
      
      logger.debug('Audit log queued for sending', { action: context.action });
    } catch (error) {
      logger.error('Failed to send audit log', {
        error: error instanceof Error ? error.message : String(error),
        context,
      });
      throw error;
    }
  }
}

// Export singleton instance
export const auditLogger = AuditLogger.getInstance();

/**
 * React Hook für Audit-Logging
 */
export function useAuditLogger() {
  return {
    log: auditLogger.log.bind(auditLogger),
    logSuccess: auditLogger.logSuccess.bind(auditLogger),
    logError: auditLogger.logError.bind(auditLogger),
    logSecurity: auditLogger.logSecurity.bind(auditLogger),
    logDataChange: auditLogger.logDataChange.bind(auditLogger),
  };
}

/**
 * Higher-Order Component für automatisches Audit-Logging
 */
export function withAuditLogging<T extends Record<string, any>>(
  Component: React.ComponentType<T>,
  defaultContext: Partial<AuditContext>,
): React.ComponentType<T> {
  return (props: T) => {
    const audit = useAuditLogger();

    // Log component mount
    React.useEffect(() => {
      audit.log({
        action: `view-${defaultContext.resource || 'component'}`,
        resource: defaultContext.resource || 'component',
        actionType: AuditActionType.READ,
        severity: AuditSeverity.LOW,
        ...defaultContext,
      });
    }, [audit]);

    return React.createElement(Component, { ...props, audit } as T);
  };
}