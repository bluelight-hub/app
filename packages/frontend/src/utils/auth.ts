import type { UserResponseDtoRoleEnum } from '@bluelight-hub/shared/client';

export const isAdmin = (role: UserResponseDtoRoleEnum | undefined): boolean => {
  return role === 'ADMIN' || role === 'SUPER_ADMIN';
};
