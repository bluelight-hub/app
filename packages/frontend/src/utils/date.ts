import dayjs, { Dayjs } from 'dayjs';
import 'dayjs/locale/de';
import customParseFormat from 'dayjs/plugin/customParseFormat';

// Konfiguriere dayjs
dayjs.extend(customParseFormat);
dayjs.locale('de');

/**
 * NATO-Datumsformate
 * NATO: ddHHmmLLLyy (Tag, Stunde, Minute, Monat, Jahr)
 * ANT: DDHHmmMMMYY (Ant Design spezifisches Format)
 */
export const DATE_FORMATS = {
    NATO: 'ddHHmmLLLyy',
    NATO_ANT: 'DDHHmmMMMYY',
} as const;

/**
 * Formatiert ein Datum in das NATO-Format
 * 
 * @param dateTime - Das zu formatierende Datum
 * @param formatType - Das zu verwendende Format (NATO oder NATO_ANT)
 * @returns Formatierter String oder undefined wenn kein Datum übergeben wurde
 */
export function formatNatoDateTime(
    dateTime: Date | Dayjs | number | string | undefined,
    formatType: keyof typeof DATE_FORMATS = 'NATO'
): string | undefined {
    if (!dateTime) {
        return undefined;
    }

    try {
        // Konvertiere zu Dayjs wenn nötig
        const date = dayjs.isDayjs(dateTime) ? dateTime : dayjs(dateTime);

        if (!date.isValid()) {
            throw new Error('Ungültiges Datum');
        }

        return date.format(DATE_FORMATS[formatType]);
    } catch (error) {
        console.error('Fehler beim Formatieren des Datums:', error);
        return undefined;
    }
}

/**
 * Parst ein Datum im NATO-Format oder Standard-Format in ein Dayjs-Objekt
 *
 * @param input - Das zu parsende Datum als String
 * @returns Dayjs Objekt
 * @throws Error wenn das Datum nicht geparst werden konnte
 */
export function parseNatoDateTime(input: string | undefined): Dayjs {
    if (!input) {
        throw new Error('Kein Datum übergeben');
    }

    try {
        // Versuche zuerst im NATO-Format zu parsen
        const natoDate = dayjs(input, DATE_FORMATS.NATO, true);
        if (natoDate.isValid()) {
            return natoDate;
        }

        // Versuche im ANT-Format zu parsen
        const antDate = dayjs(input, DATE_FORMATS.NATO_ANT, true);
        if (antDate.isValid()) {
            return antDate;
        }

        // Versuche als normales Datum zu parsen
        const fallbackDate = dayjs(input);
        if (fallbackDate.isValid()) {
            return fallbackDate;
        }

        throw new Error('Datum konnte nicht geparst werden');
    } catch (error) {
        console.error('Fehler beim Parsen des Datums:', error);
        throw new Error(`Ungültiges Datumsformat: ${input}`);
    }
}

// Für die Verwendung in Ant Design Komponenten
export const natoDateTimeAnt = DATE_FORMATS.NATO_ANT; 