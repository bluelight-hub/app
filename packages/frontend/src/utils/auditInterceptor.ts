import { logger } from './logger';
import { auditLogger } from './audit';
import {
  CreateAuditLogDtoActionTypeEnum as AuditActionType,
  CreateAuditLogDtoSeverityEnum as AuditSeverity,
} from '@bluelight-hub/shared/client';

/**
 * Bestimmt den Aktionstyp basierend auf der HTTP-Methode
 */
function getActionTypeFromMethod(method: string): AuditActionType {
  switch (method.toUpperCase()) {
    case 'GET':
      return AuditActionType.Read;
    case 'POST':
      return AuditActionType.Create;
    case 'PUT':
    case 'PATCH':
      return AuditActionType.Update;
    case 'DELETE':
      return AuditActionType.Delete;
    default:
      return AuditActionType.Read;
  }
}

/**
 * Extrahiert die Ressource aus der URL
 */
function extractResourceFromUrl(url: string): { resource: string; resourceId?: string } {
  try {
    const urlObj = new URL(url);
    const pathSegments = urlObj.pathname.split('/').filter(Boolean);

    // Remove 'api' prefix if present
    if (pathSegments[0] === 'api') {
      pathSegments.shift();
    }

    // Extract resource and potential ID
    const resource = pathSegments[0] || 'unknown';
    const resourceId =
      pathSegments.length > 1 && !pathSegments[1].includes('?') ? pathSegments[1] : undefined;

    return { resource, resourceId };
  } catch {
    return { resource: 'unknown' };
  }
}

/**
 * Bestimmt den Schweregrad basierend auf der Ressource und Aktion
 */
function determineSeverity(resource: string, method: string): AuditSeverity {
  const criticalResources = ['auth', 'users', 'roles', 'permissions'];
  const highSeverityMethods = ['DELETE', 'PUT', 'PATCH'];

  if (resource === 'auth') {
    return AuditSeverity.High;
  }

  if (criticalResources.includes(resource) && highSeverityMethods.includes(method.toUpperCase())) {
    return AuditSeverity.High;
  }

  if (method.toUpperCase() === 'DELETE') {
    return AuditSeverity.Medium;
  }

  return AuditSeverity.Low;
}

/**
 * Audit-Logging-Middleware fÃ¼r API-Aufrufe
 */
export const auditInterceptorMiddleware = {
  pre: async (context: { url: string; init: RequestInit }) => {
    // Log the start of the request
    const { resource, resourceId } = extractResourceFromUrl(context.url);
    const method = context.init.method || 'GET';

    logger.debug('API request initiated', {
      url: context.url,
      method,
      resource,
      resourceId,
    });

    // Store request start time for duration calculation
    (context as any).__auditStartTime = Date.now();
    (context as any).__auditResource = resource;
    (context as any).__auditResourceId = resourceId;

    return context;
  },

  post: async (context: { url: string; init: RequestInit; response: Response }) => {
    const { response } = context;
    const method = context.init.method || 'GET';
    const resource = (context as any).__auditResource || 'unknown';
    const resourceId = (context as any).__auditResourceId;
    const startTime = (context as any).__auditStartTime || Date.now();
    const duration = Date.now() - startTime;

    const actionType = getActionTypeFromMethod(method);
    const severity = determineSeverity(resource, method);

    // Prepare audit context
    const auditContext = {
      action: `${method.toLowerCase()}-${resource}`,
      resource,
      resourceId,
      actionType,
      severity,
      httpMethod: method,
      httpPath: context.url.startsWith('http') ? new URL(context.url).pathname : context.url,
      metadata: {
        duration,
        statusCode: response.status,
        statusText: response.statusText,
      },
    };

    if (response.ok) {
      // Log successful request
      await auditLogger.log({
        ...auditContext,
        metadata: {
          ...auditContext.metadata,
          success: true,
        },
      });
    } else {
      // Log failed request
      let errorMessage = `HTTP ${response.status}: ${response.statusText}`;

      // Try to extract error message from response
      try {
        const contentType = response.headers.get('content-type');
        if (contentType && contentType.includes('application/json')) {
          const responseClone = response.clone();
          const errorData = await responseClone.json();
          errorMessage = errorData.message || errorData.error || errorMessage;
        }
      } catch {
        // Ignore parsing errors
      }

      await auditLogger.logError(auditContext.action, resource, errorMessage, {
        ...auditContext,
        metadata: {
          ...auditContext.metadata,
          success: false,
        },
      });
    }

    return context;
  },
};

/**
 * Kombiniert Auth- und Audit-Interceptoren
 */
export function combineInterceptors(...interceptors: any[]) {
  return {
    pre: async (context: any) => {
      let result = context;
      for (const interceptor of interceptors) {
        if (interceptor.pre) {
          result = await interceptor.pre(result);
        }
      }
      return result;
    },
    post: async (context: any) => {
      let result = context;
      for (const interceptor of interceptors) {
        if (interceptor.post) {
          result = await interceptor.post(result);
        }
      }
      return result;
    },
  };
}
