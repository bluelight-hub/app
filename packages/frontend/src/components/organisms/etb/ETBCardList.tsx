import { formatNatoDateTime } from '@/utils/date';
import { logger } from '@/utils/logger';
import { EtbEntryDto, EtbEntryDtoStatusEnum, EtbKategorie } from '@bluelight-hub/shared/client';
import { Button, Empty, Tag, Tooltip } from 'antd';
import { PiEmpty, PiSwap, PiTextStrikethrough } from 'react-icons/pi';

/**
 * Props für die ETBCardList-Komponente
 */
interface ETBCardListProps {
    /**
     * Die anzuzeigenden Einsatztagebuch-Einträge
     */
    entries: EtbEntryDto[];

    /**
     * Callback für das Bearbeiten eines Eintrags
     */
    onEditEntry?: (entry: EtbEntryDto) => void;

    /**
     * Callback für das Archivieren eines Eintrags
     */
    onArchiveEntry?: (nummer: number) => void;

    /**
     * Callback für das Überschreiben eines Eintrags
     */
    onUeberschreibeEntry?: (entry: EtbEntryDto) => void;
}

/**
 * Hilfsfunktion um EtbKategorie in lesbaren Text zu konvertieren
 */
const getKategorieDisplayName = (kategorie: EtbKategorie): string => {
    const kategorieMap: Record<EtbKategorie, string> = {
        [EtbKategorie.Lagemeldung]: 'Lagemeldung',
        [EtbKategorie.Meldung]: 'Meldung',
        [EtbKategorie.Anforderung]: 'Anforderung',
        [EtbKategorie.Korrektur]: 'Korrektur',
        [EtbKategorie.AutoKraefte]: 'Auto Kräfte',
        [EtbKategorie.AutoPatienten]: 'Auto Patienten',
        [EtbKategorie.AutoTechnisch]: 'Auto Technisch',
        [EtbKategorie.AutoSonstiges]: 'Auto Sonstiges',
    };
    return kategorieMap[kategorie] || kategorie;
};

/**
 * Karten-Liste für die Mobile-Ansicht des Einsatztagebuchs
 */
export const ETBCardList: React.FC<ETBCardListProps> = ({
    entries,
    onEditEntry,
    onArchiveEntry,
    onUeberschreibeEntry,
}) => {
    if (!entries.length) {
        return <Empty image={<PiEmpty size={48} />} description="Keine Einträge verfügbar" />;
    }

    return (
        <div className="flex flex-col gap-4">
            {entries.map((entry) => {
                // Verwende die echten Properties aus EtbEntryDto
                const entryId = entry.id;
                const entryNumber = entry.laufendeNummer;
                const entryType = getKategorieDisplayName(entry.kategorie);
                const entryContent = entry.inhalt;
                const entrySender = entry.sender || "Unbekannt";
                const entryReceiver = entry.receiver;
                const isArchived = entry.istAbgeschlossen;
                const isUeberschrieben = entry.status === EtbEntryDtoStatusEnum.Ueberschrieben;

                // Datum formatieren - verwende timestampEreignis für das eigentliche Ereignis
                let formattedDate = "-";
                if (entry.timestampEreignis) {
                    try {
                        formattedDate = formatNatoDateTime(entry.timestampEreignis) ?? "-";
                    } catch (error) {
                        logger.error("Fehler beim Formatieren des Datums:", error);
                    }
                }

                return (
                    <div
                        key={entryId}
                        className={`p-3 border rounded shadow-sm ${isUeberschrieben ? 'opacity-60' : ''}`}
                    >
                        <div className="font-semibold flex justify-between mb-1 text-primary-600 dark:text-primary-400">
                            <span className="flex items-center">
                                #{entryNumber} | {entryType}
                                {isUeberschrieben && (
                                    <Tag color="warning" className="ml-2">Überschrieben</Tag>
                                )}
                            </span>
                            <span>{formattedDate}</span>
                        </div>
                        <p className={`text-sm ${isUeberschrieben ? 'line-through' : ''}`}>{entryContent}</p>
                        <div className="text-xs mt-1 text-gray-500 flex items-center justify-between">
                            <span>Absender: {entrySender}</span>
                            <span>Empfänger: {entryReceiver}</span>
                        </div>
                        {!isArchived && !isUeberschrieben && (
                            <div className="mt-2 flex gap-2 justify-end">
                                <Tooltip title="Eintrag überschreiben">
                                    <Button
                                        size="small"
                                        data-testid="ueberschreiben-button"
                                        onClick={() => onUeberschreibeEntry && onUeberschreibeEntry(entry)}
                                        icon={<PiSwap />}
                                    />
                                </Tooltip>
                                <Tooltip title="Eintrag bearbeiten">
                                    <Button
                                        size="small"
                                        data-testid="edit-button"
                                        onClick={() => onEditEntry && onEditEntry(entry)}
                                        icon={<PiSwap />}
                                    />
                                </Tooltip>
                                <Tooltip title="Eintrag streichen">
                                    <Button
                                        size="small"
                                        data-testid="archive-button"
                                        onClick={() => onArchiveEntry && onArchiveEntry(entryNumber)}
                                        danger
                                        icon={<PiTextStrikethrough />}
                                    />
                                </Tooltip>
                            </div>
                        )}
                    </div>
                );
            })}
        </div>
    );
};
