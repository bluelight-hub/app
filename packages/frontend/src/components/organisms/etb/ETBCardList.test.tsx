import * as dateUtils from '@/utils/date';
import { logger } from '@/utils/logger';
import { EtbEntryDto, EtbEntryDtoStatusEnum } from '@bluelight-hub/shared/client';
import { fireEvent, render, screen } from '@testing-library/react';
import { MockedFunction, vi } from 'vitest';
import { ETBCardList } from './ETBCardList';

// Der ETBEntryFields Typ aus der Komponente
type ETBEntryFields = {
    nummer?: number;
    type?: string;
    timestamp?: Date;
    content?: string;
    sender?: string;
    receiver?: string;
    archived?: boolean;
    status?: 'aktiv' | 'ueberschrieben';
};

// Mock der logger-Funktionen
vi.mock('@/utils/logger', () => ({
    logger: {
        error: vi.fn(),
        info: vi.fn(),
    }
}));

// Mock für die formatNatoDateTime-Funktion
vi.mock('@/utils/date', () => ({
    formatNatoDateTime: vi.fn()
}));

describe('ETBCardList', () => {
    const mockOnEditEntry = vi.fn();
    const mockOnArchiveEntry = vi.fn();
    const mockOnUeberschreibeEntry = vi.fn();

    // Fixtures - mit zusätzlichen Feldern, die vom ETBEntryFields-Interface erwartet werden
    const createMockEntry = (overrides = {}): EtbEntryDto & ETBEntryFields => ({
        id: 'test-id-1',
        laufendeNummer: 1,
        timestampErstellung: new Date('2023-01-01T10:00:00Z'),
        timestampEreignis: new Date('2023-01-01T09:30:00Z'),
        autorId: 'author-1',
        autorName: 'Max Mustermann',
        autorRolle: 'Einsatzleiter',
        kategorie: 'Standard',
        titel: 'Test Eintrag',
        beschreibung: 'Dies ist ein Testeintrag',
        referenzEinsatzId: 'einsatz-1',
        referenzPatientId: null,
        referenzEinsatzmittelId: null,
        systemQuelle: 'TEST',
        version: 1,
        istAbgeschlossen: false,
        timestampAbschluss: null,
        abgeschlossenVon: null,
        status: EtbEntryDtoStatusEnum.Aktiv,
        // ETBEntryFields zusätzliche Felder
        nummer: 1,
        type: 'Standard',
        timestamp: new Date('2023-01-01T09:30:00Z'),
        content: 'Dies ist ein Testeintrag',
        sender: 'Max Mustermann',
        receiver: 'Einsatzleitung',
        archived: false,
        ...overrides
    });

    beforeEach(() => {
        vi.resetAllMocks();
        (dateUtils.formatNatoDateTime as MockedFunction<typeof dateUtils.formatNatoDateTime>).mockImplementation(() => '01100023jan23');
    });

    test('sollte "Keine Einträge verfügbar" anzeigen, wenn die Liste leer ist', () => {
        render(<ETBCardList entries={[]} />);

        expect(screen.getByText('Keine Einträge verfügbar')).toBeInTheDocument();
    });

    test('sollte alle Einträge korrekt anzeigen', () => {
        const entries = [
            createMockEntry({ id: 'test-id-1', nummer: 1, laufendeNummer: 1 }),
            createMockEntry({ id: 'test-id-2', nummer: 2, laufendeNummer: 2, content: 'Zweiter Eintrag', beschreibung: 'Zweiter Eintrag' })
        ];

        render(<ETBCardList entries={entries} />);

        // Text ist aufgeteilt im DOM, deshalb prüfen wir spezifische Texte
        const nummerElements = screen.getAllByText(/#\d+/);
        expect(nummerElements.length).toBe(2);
        expect(nummerElements[0].textContent).toContain('#1');
        expect(nummerElements[1].textContent).toContain('#2');

        // Prüfen, ob beide Standardtypen vorhanden sind
        const standardElements = screen.getAllByText(/Standard/);
        expect(standardElements.length).toBe(2);

        // Prüfen, ob die Inhalte korrekt sind
        expect(screen.getByText('Dies ist ein Testeintrag')).toBeInTheDocument();
        expect(screen.getByText('Zweiter Eintrag')).toBeInTheDocument();
    });

    test('sollte Einträge mit fehlenden Feldern korrekt anzeigen', () => {
        // Eintrag mit fehlenden Standardfeldern
        const entryWithMissingFields = {
            id: 'missing-fields',
            content: 'Inhalt ohne vollständige Struktur'
        } as unknown as EtbEntryDto;

        render(<ETBCardList entries={[entryWithMissingFields]} />);

        // Prüfen, ob der Inhalt korrekt angezeigt wird
        expect(screen.getByText('Inhalt ohne vollständige Struktur')).toBeInTheDocument();

        // Prüfen, ob die Standardtexte vorhanden sind
        expect(screen.getByText('Absender:')).toBeInTheDocument();
        expect(screen.getByText('Empfänger:')).toBeInTheDocument();
    });

    test('sollte überschriebene Einträge korrekt anzeigen', () => {
        const überschriebenerEintrag = createMockEntry({
            status: EtbEntryDtoStatusEnum.Ueberschrieben,
            content: 'Überschriebener Eintrag'
        });

        render(<ETBCardList entries={[überschriebenerEintrag]} />);

        expect(screen.getByText('Überschrieben')).toBeInTheDocument();
        // Prüfen, ob es ein Element mit der CSS-Klasse line-through gibt
        const überschriebenElement = screen.getByText('Überschriebener Eintrag');
        expect(überschriebenElement.className).toContain('line-through');
    });

    test('sollte Fehler bei der Datumsformatierung behandeln', () => {
        // Logger-Mock für diesen Test explizit überwachen
        const errorSpy = vi.spyOn(logger, 'error');

        // Simulieren eines Fehlers bei der Datumsformatierung
        (dateUtils.formatNatoDateTime as MockedFunction<typeof dateUtils.formatNatoDateTime>).mockImplementation(() => {
            throw new Error('Formatierungsfehler');
        });

        render(<ETBCardList entries={[createMockEntry()]} />);

        // Prüfen, ob der Logger den Fehler korrekt protokolliert hat
        expect(errorSpy).toHaveBeenCalledWith(
            'Fehler beim Formatieren des Datums:',
            expect.any(Error)
        );

        // Prüfen, ob der Fallback-Wert angezeigt wird
        expect(screen.getAllByText('-').length).toBeGreaterThan(0);
    });

    test('sollte den onEditEntry-Callback korrekt aufrufen', () => {
        const entry = createMockEntry();
        render(
            <ETBCardList
                entries={[entry]}
                onEditEntry={mockOnEditEntry} />
        );

        // Alle Buttons finden und nach Icon-Inhalt identifizieren
        const buttons = screen.getAllByRole('button');

        // Klicke auf jeden Button und prüfe, welcher den Callback auslöst
        for (const button of buttons) {
            fireEvent.click(button);
        }

        // Prüfen, ob der Callback mindestens einmal aufgerufen wurde
        expect(mockOnEditEntry).toHaveBeenCalledWith(entry);
    });

    test('sollte den onArchiveEntry-Callback korrekt aufrufen', () => {
        const entry = createMockEntry({
            nummer: 42,
            laufendeNummer: 42
        });

        render(
            <ETBCardList
                entries={[entry]}
                onArchiveEntry={mockOnArchiveEntry}
            />
        );

        // Alle Buttons finden und klicken
        const buttons = screen.getAllByRole('button');

        // Klicke auf jeden Button und prüfe, welcher den Callback auslöst
        for (const button of buttons) {
            fireEvent.click(button);
        }

        // Prüfen, ob der Callback mindestens einmal aufgerufen wurde
        expect(mockOnArchiveEntry).toHaveBeenCalledWith(42);
    });

    test('sollte den onUeberschreibeEntry-Callback korrekt aufrufen', () => {
        const entry = createMockEntry();
        render(
            <ETBCardList
                entries={[entry]}
                onUeberschreibeEntry={mockOnUeberschreibeEntry}
            />
        );

        // Alle Buttons finden und klicken
        const buttons = screen.getAllByRole('button');

        // Klicke auf jeden Button und prüfe, welcher den Callback auslöst
        for (const button of buttons) {
            fireEvent.click(button);
        }

        // Prüfen, ob der Callback mindestens einmal aufgerufen wurde
        expect(mockOnUeberschreibeEntry).toHaveBeenCalledWith(entry);
    });

    test('sollte keine Aktions-Buttons für archivierte Einträge anzeigen', () => {
        const archivedEntry = createMockEntry({
            archived: true
        });

        render(
            <ETBCardList
                entries={[archivedEntry]}
                onEditEntry={mockOnEditEntry}
                onArchiveEntry={mockOnArchiveEntry}
                onUeberschreibeEntry={mockOnUeberschreibeEntry}
            />
        );

        // Es sollten keine Button-Elemente vorhanden sein
        expect(screen.queryAllByRole('button').length).toBe(0);
    });

    test('sollte keine Aktions-Buttons für überschriebene Einträge anzeigen', () => {
        const überschriebenerEintrag = createMockEntry({
            status: EtbEntryDtoStatusEnum.Ueberschrieben
        });

        render(
            <ETBCardList
                entries={[überschriebenerEintrag]}
                onEditEntry={mockOnEditEntry}
                onArchiveEntry={mockOnArchiveEntry}
                onUeberschreibeEntry={mockOnUeberschreibeEntry}
            />
        );

        // Es sollten keine Button-Elemente vorhanden sein
        expect(screen.queryAllByRole('button').length).toBe(0);
    });
}); 