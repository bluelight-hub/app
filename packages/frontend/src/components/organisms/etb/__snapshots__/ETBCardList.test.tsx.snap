// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ETBCardList > sollte mit Standard-Props dem Snapshot entsprechen 1`] = `
<div>
  <div
    class="flex flex-col gap-4"
  >
    <div
      class="p-3 border rounded shadow-sm "
    >
      <div
        class="font-semibold flex justify-between mb-1 text-primary-600 dark:text-primary-400"
      >
        <span
          class="flex items-center"
        >
          #
          1
           | 
          Meldung
        </span>
        <span>
          01100023jan23
        </span>
      </div>
      <p
        class="text-sm "
      >
        Dies ist ein Testeintrag
      </p>
      <div
        class="text-xs mt-1 text-gray-500 flex items-center justify-between"
      >
        <span>
          Absender: 
          Max Mustermann
        </span>
        <span>
          Empf채nger: 
          Einsatzleitung
        </span>
      </div>
      <div
        class="mt-2 flex gap-2 justify-end"
      >
        <div
          data-testid="mock-tooltip"
          data-title="Eintrag 체berschreiben"
        >
          <button
            data-testid="ueberschreiben-button"
          >
            Icon
          </button>
        </div>
        <div
          data-testid="mock-tooltip"
          data-title="Eintrag bearbeiten"
        >
          <button
            data-testid="edit-button"
          >
            Icon
          </button>
        </div>
        <div
          data-testid="mock-tooltip"
          data-title="Eintrag streichen"
        >
          <button
            data-testid="archive-button"
          >
            Icon
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`ETBCardList > sollte mit leeren Daten dem Snapshot entsprechen 1`] = `
<div>
  <div
    data-testid="mock-empty"
  >
    Keine Eintr채ge verf체gbar
  </div>
</div>
`;
